File: SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.m
%% 資料夾路徑
tic
%import SMSPD_waveform_plot_ChatGPT.*
folder_path = 'D:\game\Pulse_800_0nW_0degrees';

% 實驗參數
% Exp_para = 'SMSPD_NbTiN_1_1-1_Pulse_450_30000nW_0degrees_';

% 電壓範圍'
file_list = dir(fullfile(folder_path, '*mV.txt'));

if (isempty(file_list))
    error('No text file is found!')
end
[basename,~, ~] = extract_info(file_list(1).name);

% sort by voltage
%Vb = sort(extract_mV_from_list(file_list));
[Vb, Ib] = extract_mV_from_list(file_list);
file_table = struct2table(file_list);
file_table = addvars(file_table,Vb, Ib,'Before',1);
file_table = sortrows(file_table,1);

%Vb = [5,10:20,25,30];
%Vb = [30,25];

%% 預分配效率數組

% threshold
STDEV_CUT = 0.05; % Threshold of STDEV 標準差閾值
V_CUT = 0.03; % Threshold of voltage amplitude 閾值
index_setting = 1; % 1: 800nm, 2: 515nm


Wavelength = ['800 nm'; '515 nm'];
datalen = [125; 250];
nPulse = [4; 1];
laserConf = table(Index, datalen, nPulse);

Nevent = 10000; % 1 ~ 10000
DATA_LENGTH = laserConf.datalen(index_setting); % The number of data points of each event 每個事件的數據點數目
NUM_PEAKS = laserConf.nPulse(index_setting); % The number of signal peak in each event (usually = 1) 一個事件有幾個peak，通常是1

CONTROL_REGION = [1 15]; % The range of control region 沒有訊號的數據點
%CONTROL_REGION = [1 DATA_LENGTH];  % defaul setting 預設設定
PEAK_LENGTH = ceil(DATA_LENGTH/NUM_PEAKS); % The range of first peak 第一個peak的數據點數目

eff = zeros(length(Vb), 1);
temp_waveform = NaN(DATA_LENGTH,50);
VmaxArray = zeros(Nevent, 50);
VmaxIndexArray = zeros(Nevent, 50);

% setting for tab with figures
fig = uifigure('Name', 'Multi-Tab Plots', 'Position',[40 80 1400 700]);
tabgroup = uitabgroup(fig, 'Position', [20 20 1300 650]);

temp_j = 1;
% for loop for different Ib, Vb files
for k = 1:length(file_table.Vb)
%for k = 6:12
    % d = load([Exp_para, num2str(Vb(k)), '_mV.txt']);
    % file_path = fullfile(folder_path,file_list(k).name);
    file_path = fullfile(folder_path,string(file_table(1,:).name)); 
    disp(['processing... ', num2str(k), '/',num2str(length(file_table.Vb))])
    
    d = load(file_path, '-ascii'); % Loading data
    signal = d(:, 1);
    trigger = d(:, 2);
    clear d

    if size(signal) ~= (Nevent+1)*DATA_LENGTH
        warning('k = %d. Data point does no match the size', k)
    end
    
    q = zeros(PEAK_LENGTH, 1);
    nPass = 0; nFail_pre = 0; nFail = 0;
    Vamplitude = zeros(Nevent, 1);
    jitter = NaN(Nevent, 1);
    sigma = zeros(Nevent, 1);
    %Vmax = zeros(Nevent, 1);
    %VmaxIndex = zeros(Nevent, 1);
    

    Raw_sig_ave = zeros(DATA_LENGTH,1);
    temp_sig = zeros(DATA_LENGTH,1);
    fail_presel = zeros(PEAK_LENGTH, 1);
    fail_sel = zeros(PEAK_LENGTH, 1);
    sig_region = zeros(PEAK_LENGTH, 1);
    
    warning_preselction = true;
    
    % loop Events in one of files
    for i = 1:Nevent 
        
        index_peak = (1:PEAK_LENGTH) + DATA_LENGTH * (i); % index of i events
        index = (1:DATA_LENGTH) + DATA_LENGTH * (i); % index of i events
        Raw_sig_ave = Raw_sig_ave + signal(index);
        s = signal(index_peak);  % signal of 1st peak in i event, length = DATA_LENGTH
        s_raw = signal(index);   % signal of all peaks in i event, length = DATA_LENGTH

        % Calculate sigma 計算標準差
        sigma(i) = std(s(1:PEAK_LENGTH));   % the range can be changed
        %[Vmax(i), VmaxIndex(i)] = max(s);
        [VmaxArray(i,k), VmaxIndexArray(i,k)] = max(s);

        if (warning_preselction && nFail_pre > Nevent*0.1)
            warning_preselction = false;
            warning('k = %d. Too many events fail pre-selection', k)
        end

        %if (true && k>=5 && k<=10 && i==100)
        if (true  && i==100)
            temp_waveform(:,temp_j) = s_raw;
            %outputname = [basename, '_test_waveform_',num2str(temp_j), '.txt'];
            %save(fullfile(folder_path, outputname), 's_raw', '-ascii');
            temp_j = temp_j + 1;
        end
        if (i==100)
            temp_sig = s_raw;
        end
        % pre-selection
        if true & sigma(i) <= STDEV_CUT
            q = s + q;
            s1 = s(1:PEAK_LENGTH);  % The data point of signal region
            %sbg = s(CONTROL_REGION); % control region
            %Vamplitude(i) = max(s1) - min(s1); 
            Vamplitude(i) = max(s1) - min(s1(CONTROL_REGION)); % signal amplitude
            deltaSig = diff(s1);  % i+1 data point - i data point
            dtr = diff(trigger(index_peak));
            ndeltaSig = find(deltaSig == max(deltaSig), 1);
            ntr = find(dtr == max(dtr), 1);
            jitter(i) = ntr(1) - ndeltaSig(1);
            % selection
            count = length(find(s1 > V_CUT)); % selection (Voltage cut)
            if count >= 2
                sig_region = sig_region + s1;
                count = 1;
                nPass = nPass + 1;
            else
                fail_sel = fail_sel + s1;
                nFail = nFail + 1;
            end
            % nPass = nPass + count;
       else 
           fail_presel = fail_presel + s;
           nFail_pre = nFail_pre + 1;
       end % end of pre-selection
    end 
    % finish loop over event in one of files

    fail_presel = fail_presel/nFail_pre;
    fail_sel = fail_sel/nFail;
    sig_region = sig_region/nPass;
    Raw_sig_ave = Raw_sig_ave/Nevent;


    % Calculate the efficiency 計算效率
    %Ef_event = length(find(sigma <= STDEV_CUT)); 
    Ef_event = nPass + nFail;
    eff(k) = nPass / Ef_event;
    
    % The configure of plots 顯示圖形
    %figure;
    %tab(k)=uitab(tabgroup,'Title', sprintf('Tab_%i', k));
    tab(k)=uitab(tabgroup,'Title', sprintf('%.0f uA', file_table(k,:).Ib));
    t = tiledlayout(tab(k), 3, 3, 'Padding', 'compact', 'TileSpacing', 'compact');
    %tab_axes = axes('parent',tab(k));
    %hold(tab_axes,'on')
    
    %subplot(2, 2, 1);
    tile1 = nexttile(t, 1);
    plot(tile1,sig_region, 'g');
    title(tile1,'Signal-ave');

    tile2 = nexttile(t);
    plot(tile2,fail_sel, 'g');
    title(tile2,'fail-sel-ave');

    tile3 = nexttile(t);
    plot(tile3,fail_presel, 'g');
    title(tile3,'fail-presel-ave');

    tile4 = nexttile(t);
    plot(tile4,Raw_sig_ave, 'g');
    ylim(tile4,[min(-0.01,min(Raw_sig_ave)*1.1), max(0.01,max(Raw_sig_ave)*1.1)]);
    title(tile4,'Raw-Data-ave');

    %subplot(2, 2, 2);
    tile5 = nexttile(t);
    %histogram(tile5,Vmax);
    histogram(tile5,VmaxArray(:,k));
    %xlim(tile5,[-0.01, max(0.01,max(Vmax)*1.1)]);
    xlim(tile5,[-0.01, max(0.01,max(VmaxArray(:,k))*1.1)]);
    title(tile5,'Histogram of Vmax');

    %subplot(2, 2, 4);
    tile6 = nexttile(t);
    %plot(tile6,Vamplitude);
    %histogram(tile6,VmaxIndex);
    histogram(tile6,VmaxIndexArray(:,k));
    xlim(tile6,[1, PEAK_LENGTH]);
    title(tile6,'Histogram of VmaxIndex');

    %subplot(2, 2, 3);
    tile7 = nexttile(t);
    plot(tile7,temp_sig, 'g');
    ylim(tile7,[min(-0.01,min(temp_sig)*1.1), max(0.01,max(temp_sig)*1.1)]);
    title(tile7,'100th event waveform');
    %plot(tile7,deltaSig, 'g');
    %title(tile7,'deltaSig');


    
    % Summary statistic table
    vars = ["File Name";"nPass";"nFail";"nFail_pre";"Effi";"Total"];
    passEve = [string(file_table(1,:).name);nPass;nFail;nFail_pre;eff(k);nPass+nFail+nFail_pre];
    Config_name = ["STDEV-CUT";"V-CUT";"DATA-LENGTH";"PEAK-LENGTH";"CONTROAL-REGION";" "];
    Config_cut = [STDEV_CUT;V_CUT;DATA_LENGTH;PEAK_LENGTH;strjoin(string(CONTROL_REGION));" "];
    tdata = table(vars,passEve,Config_name,Config_cut,'VariableNames',{'Variable','Name/ # of Events','Config_name','Value'});
    uit = uitable(tab(k),"Data",tdata,'Units', 'Normalized','Position', [0.5 0.0 0.4 0.2]);

end 
% end of loop files

% summary effi plot
tab(k+1)=uitab(tabgroup,'Title', sprintf('Effi'));
ax = uiaxes(tab(k+1), 'Position', [40 40 1200 500]);
%plot(ax,file_table.Vb,eff)
plot(ax,file_table.Ib,eff)
title(ax, 'Efficiency vs Bias Current');
%xlabel(ax, 'Bias Voltage (mV)');
xlabel(ax, 'Bias Current (mA)');
ylabel(ax, 'Efficiency');


% 將Efficiency保存到 txt 檔案

F = [file_table.Vb, eff];
%outputname = [basename, '_',num2str(V_CUT),'_', num2str(STDEV_CUT), '_mV_efficiency.txt'];
outputname = [basename, '_',num2str(V_CUT),'_noSTDEVcut_mV_efficiency.txt'];
save(fullfile(folder_path, outputname), 'F', '-ascii');

%outputnameFig = [basename, '_',num2str(V_CUT),'_', num2str(STDEV_CUT),'.fig'];
outputnameFig = [basename, '_',num2str(V_CUT),'_noSTDEVcut.fig'];
savefig(fig,fullfile(folder_path, outputnameFig));  

outputnameVmax = [basename, '_Vmax.txt'];
save(fullfile(folder_path, outputnameVmax), 'VmaxArray', '-ascii',"-tabs");
outputnameVmax = [basename, '_VmaxIndex.txt'];
save(fullfile(folder_path, outputnameVmax), 'VmaxIndexArray', '-ascii',"-tabs");

%disp(['save data to ', fullfile(folder_path, outputname)]);
% save waveform
outputname2 = [basename, '_waveform_all.txt'];
%save(fullfile(folder_path, outputname2), "temp_waveform", '-ascii');

disp(['save data to ', folder_path]);
disp(['Text file name: ', outputname]);

%% function block
function [basename, mV_value, Ib] = extract_info(filename)
    % Extracts the mV value from a filename string.
    % Example: 'SMSPD_NbTiN_1_1-1_Pulse_450_30000nW_0degrees_100_mV.txt'
    % Output: mV_value, Ib. Type: double, double
    % basenmae: prefix
    % 

    % Regular expression to find the mV value before '_mV'

    % Apply regular expression
    tokens  = regexp(filename, '_(\d+)mV', 'tokens');
    uA      = regexp(filename, '_(\d+)uA', 'tokens');
    prefix  = regexp(filename, '^(.*?Pulse_)', 'tokens');
    Ib      = str2double(uA{1}{1});
    % Check if a match is found
    if ~isempty(tokens)
        basename = prefix{1}{1};
        mV_value = str2double(tokens{1}{1});
        %mV_value = str2double(tokens{1}{2}); % Convert extracted string to number
    else
        error(['No mV value found in filename. filename: ', filename]);
    end
end

function [mV_values, Ib] = extract_mV_from_list(filenames)
    % Extracts mV values from a cell array of filenames.
    % Input: cell array of filenames
    % Output: array of extracted mV values

    mV_values = NaN(length(filenames),1); % Preallocate array for mV values
    Ib = NaN(length(filenames),1); % Preallocate array for mV values

    parfor i = 1:length(filenames)
        [~, mV_values(i), Ib(i)] = extract_info(filenames(i,:).name);
        % mV_values(i) = extract_mV(filenames(i));
    end
end

toc

File: SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.py
"""
效率與 time jitter 分析（對應 SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.m 舊版邏輯）

加速版：用 numpy 向量化一次處理所有事件，取代逐事件的 for 迴圈。
（電壓檔之間還是序列處理，因為要逐張畫圖；速度瓶頸在每個檔內部的事件迴圈）

執行方式：
    1) 不帶參數 — 使用「設定區」的 folder_path
       python SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.py

    2) 指定資料夾
       python SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.py -d "C:\\path\\to\\folder"

    3) 指定單一 .txt 檔
       python SMSPD_ana_for_eff_time_jitter_tdms_chatGPT.py -i "C:\\path\\to\\file_mV.txt"
"""

import os
import re
import time
import argparse
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


def _fast_loadtxt(path):
    """以 pandas C engine 解析 ASCII 數值檔，比 np.loadtxt 快 5-10×"""
    return pd.read_csv(path, sep=r'\s+', header=None,
                       dtype=np.float64, engine='c').values


# =====================================================================
# 設定區 — 請依實驗環境修改
# =====================================================================

folder_path = r'E:\SMSPD_NbTiN_1\Laser\1-1\20250107\12\Pulse\450\10000kHz\15000nW\0degrees\20250107_235851\Pulse_450_15000nW_0degrees'

# 每個電壓檔分析多少事件
Nevent = 1000

# 每個事件的取樣點數
DATA_LENGTH = 1000

# 標準差閾值：基線雜訊大於此值的事件被視為無效
stdth = 0.02

# ds1 閾值（保留與原 .m 一致，目前僅作為輸出檔名用）
ds1th = 0.05


# =====================================================================
# 工具函式
# =====================================================================
def extract_info(filename):
    match = re.search(r'(.*)_(\d+)_mV\.txt', filename)
    if not match:
        raise ValueError(f'檔名格式不符: {filename}')
    return match.group(1), int(match.group(2))


# =====================================================================
# 主程式
# =====================================================================

t_start = time.time()

# --- CLI 覆寫 ---
parser = argparse.ArgumentParser(description='SMSPD efficiency / jitter analysis')
group = parser.add_mutually_exclusive_group()
group.add_argument('-i', '--input', metavar='FILE', help='單一 *_mV.txt 檔')
group.add_argument('-d', '--dir',   metavar='DIR',  help='含 *_mV.txt 的資料夾')
cli = parser.parse_args()

if cli.input:
    folder_path = os.path.dirname(os.path.abspath(cli.input))
    all_files = [os.path.basename(cli.input)]
elif cli.dir:
    folder_path = cli.dir
    all_files = [f for f in os.listdir(folder_path) if f.endswith('_mV.txt')]
else:
    all_files = [f for f in os.listdir(folder_path) if f.endswith('_mV.txt')]

if not all_files:
    raise RuntimeError('No text file is found!')

file_with_voltage = [(f, extract_info(f)[1]) for f in all_files]
file_with_voltage.sort(key=lambda x: x[1])
file_list = [f for f, _ in file_with_voltage]
Va = np.array([v for _, v in file_with_voltage])
basename = extract_info(file_list[0])[0]

# 預配置效率陣列
eff = np.zeros(len(Va))

# 對每個電壓檔做分析
for k in range(len(Va)):
    file_path = os.path.join(folder_path, file_list[k])
    print(f'processing... {k+1}/{len(Va)}')

    # 載入 ASCII 檔（兩欄：signal, trigger）
    d = _fast_loadtxt(file_path)
    signal = d[:, 0]
    trigger = d[:, 1]

    # =================================================================
    # 對齊原 .m 的索引邏輯：
    #   for i = 1..Nevent: r = (1:DATA_LENGTH) + DATA_LENGTH*i
    # 也就是說第一段資料從 index DATA_LENGTH 開始，總共取 Nevent 段
    # 若資料長度不足，自動縮減到實際可分析的事件數
    # =================================================================
    actual_Nevent = min(Nevent, len(signal) // DATA_LENGTH - 1)

    # 把訊號重塑成 2D：每一列是一個事件，共 actual_Nevent 列、DATA_LENGTH 行
    sig_mat = signal[DATA_LENGTH : DATA_LENGTH + actual_Nevent * DATA_LENGTH] \
              .reshape(actual_Nevent, DATA_LENGTH)
    trg_mat = trigger[DATA_LENGTH : DATA_LENGTH + actual_Nevent * DATA_LENGTH] \
              .reshape(actual_Nevent, DATA_LENGTH)

    # =================================================================
    # 一次計算所有事件的統計量（這就是「向量化」的核心）
    # axis=1 表示「對每列做運算」，所以結果是長度 = actual_Nevent 的陣列
    # =================================================================

    # 每個事件的標準差（ddof=1 對應 Matlab 的 std）
    sigma = np.std(sig_mat, axis=1, ddof=1)

    # 哪些事件雜訊夠小，算有效事件
    valid = sigma <= stdth

    # 每個事件的最大-最小幅度（無效事件填 NaN）
    Z = np.full(actual_Nevent, np.nan)
    Z[valid] = sig_mat[valid].max(axis=1) - sig_mat[valid].min(axis=1)

    # 訊號與觸發的差分（找上升沿位置用）
    ds = np.diff(sig_mat, axis=1)
    dtr = np.diff(trg_mat, axis=1)

    # 每個事件中「最大斜率」的取樣位置
    nds1 = np.argmax(ds, axis=1)
    ntr = np.argmax(dtr, axis=1)

    # time jitter = 觸發上升沿位置 - 訊號上升沿位置
    jitter = np.full(actual_Nevent, np.nan)
    jitter[valid] = ntr[valid] - nds1[valid]

    # =================================================================
    # 計算效率：count = 該事件有沒有偵測到光子
    # 原 .m 邏輯：count = 樣本數 (s > stdth)；若 >= 2 則 count = 1
    # 等價於：「事件中只要有任何一個樣本 > stdth，count = 1，否則 0」
    # =================================================================
    count = (sig_mat > stdth).any(axis=1).astype(int)

    # X = 偵測到光子的有效事件數
    X = int(count[valid].sum())
    Ef_event = int(valid.sum())

    if Ef_event > 0:
        eff[k] = X / Ef_event
    else:
        eff[k] = 0.0
        print(f'  [警告] 電壓 {Va[k]} 沒有有效事件')

    # =================================================================
    # 畫圖：找出最後一個有效事件當代表，畫四連格
    # =================================================================
    valid_indices = np.where(valid)[0]
    if len(valid_indices) > 0:
        last_valid = valid_indices[-1]
        s1 = sig_mat[last_valid]
        ds1 = ds[last_valid]
    else:
        s1 = np.zeros(DATA_LENGTH)
        ds1 = np.zeros(DATA_LENGTH - 1)

    fig, axes = plt.subplots(2, 2, figsize=(10, 8))
    fig.suptitle(f'Voltage = {Va[k]} mV')

    axes[0, 0].plot(s1, 'g')
    axes[0, 0].set_title('Signal s1 (last valid event)')

    axes[0, 1].hist(np.sort(sigma), bins=50)
    axes[0, 1].set_title('Histogram of Sigma')

    axes[1, 0].plot(ds1, 'g')
    axes[1, 0].set_title('ds1 = diff(s1)')

    axes[1, 1].plot(Z, '.')
    axes[1, 1].set_title('Z (max - min)')

    plt.tight_layout()
    fig_path = os.path.join(folder_path, f'{basename}_{Va[k]}mV_analysis.png')
    plt.savefig(fig_path, dpi=120, bbox_inches='tight')
    plt.close(fig)


# =====================================================================
# 輸出最終 (Voltage, efficiency) 結果
# =====================================================================
F = np.column_stack([Va, eff])
output_name = f'{basename}_{ds1th}_{stdth}_mV_efficiency.txt'
output_path = os.path.join(folder_path, output_name)
np.savetxt(output_path, F, fmt='%14.7e')
print(f'save data to {output_name}')

print(f'Done. Elapsed: {time.time() - t_start:.2f} s')
